openssl/src/kdf.rs

```diff
@@ -34,15 +34,60 @@ cfg_if::cfg_if! {
         use crate::lib_ctx::LibCtxRef;
         use crate::error::ErrorStack;
 
-        /// Derives a key using the argon2id algorithm.
+        #[allow(clippy::too_many_arguments)]
+        pub fn argon2d(
+            ctx: Option<&LibCtxRef>,
+            pass: &[u8],
+            salt: &[u8],
+            ad: Option<&[u8]>,
+            secret: Option<&[u8]>,
+            iter: u32,
+            lanes: u32,
+            memcost: u32,
+            out: &mut [u8],
+        ) -> Result<(), ErrorStack> {
+            return argon2_helper(b"ARGON2D\0", ctx, pass, salt, ad, secret, iter, lanes, memcost, out);
+        }
+
+        #[allow(clippy::too_many_arguments)]
+        pub fn argon2i(
+            ctx: Option<&LibCtxRef>,
+            pass: &[u8],
+            salt: &[u8],
+            ad: Option<&[u8]>,
+            secret: Option<&[u8]>,
+            iter: u32,
+            lanes: u32,
+            memcost: u32,
+            out: &mut [u8],
+        ) -> Result<(), ErrorStack> {
+            return argon2_helper(b"ARGON2I\0", ctx, pass, salt, ad, secret, iter, lanes, memcost, out);
+        }
+
+        #[allow(clippy::too_many_arguments)]
+        pub fn argon2id(
+            ctx: Option<&LibCtxRef>,
+            pass: &[u8],
+            salt: &[u8],
+            ad: Option<&[u8]>,
+            secret: Option<&[u8]>,
+            iter: u32,
+            lanes: u32,
+            memcost: u32,
+            out: &mut [u8],
+        ) -> Result<(), ErrorStack> {
+            return argon2_helper(b"ARGON2ID\0", ctx, pass, salt, ad, secret, iter, lanes, memcost, out);
+        }
+
+        /// Derives a key using the argon2* algorithms.
         ///
         /// To use multiple cores to process the lanes in parallel you must
         /// set a global max thread count using `OSSL_set_max_threads`. On
         /// builds with no threads all lanes will be processed sequentially.
         ///
         /// Requires OpenSSL 3.2.0 or newer.
-        #[allow(clippy::too_many_arguments)]
-        pub fn argon2id(
+        fn argon2_helper(
+            kdf_identifier: &'static [u8],
```

Comment:
for soundness this needs to be a `&CStr` or something, otherwise you get buffer overread if you pass something without a trailing nul

Reply:
The `c""` syntax is not compatible with our MSRV